bulletin-buddy-post-service/Dockerfile

```diff
@@ -4,10 +4,17 @@
 # Run:
 # $> docker run -p 8081:8081 -i -t bulletin-buddy/backend:latest
 
-FROM openjdk:17-jdk-alpine
+FROM maven:3-eclipse-temurin-17-alpine as maven
+
+WORKDIR /app
+COPY . /app
 
-COPY ./target/bulletin-buddy-post-service-0.0.1-SNAPSHOT.jar /spring-boot.jar
+RUN mvn package
+
+FROM openjdk:17-jdk-alpine
 
+WORKDIR /app
+COPY --from=maven /app/target/bulletin-buddy-post-service-*.*.*.jar /app/spring-boot.jar
```

Comment:
When I run:
```
bullettin-buddy/bulletin-buddy-post-service$> docker build -t post-service .
```
I get this error:
```
...
Step 7/9 : COPY --from=maven /app/target/bulletin-buddy-post-service-*.*.*.jar /app/spring-boot.jar
When using COPY with more than one source file, the destination must be a directory and end with a /
```
The reason for this is, that `*.*.*.jar` can theoretically match multiple files, but you want to much these multiple files into a single `spring-boot.jar`. That's why the docker is complaining and won't build.

I would suggest to rollback to the previous way, where we specifically told which file we want, i.e. use this instead:
```
COPY --from=maven /app/target/bulletin-buddy-post-service-bulletin-buddy-post-service-0.0.1-SNAPSHOT.jar /app/spring-boot.jar
```
Or do you have another ideas?

Reply:
I'm trying to replicate the error but it works fine on both my machine and on the pipelines... I suppose you get the problem because you have to rebuild the application to generate a fresh .jar with the new version (in this case the issue would be caused by the '-SNAPSHOT' after the version that doesn't get recognized by the pseudo-regex I assumo).

The two easiest solutions would be:

1. as you said, revert the the explicit version, but note that this would require to both change manually the modules version AND the Dockerfiles copy path
2. investigate if some more complex expression can parse every kind of version we could throw at it

Reply:
Another thing that comes to mind would be to automatically update and change the version in the Dockerfile by picking up the modules versions inside the CI/CD action... I think it can be done but I cannot guarantee it right now. I think we can stick to one of the two previously proposed solutions for now.

We could also do some really nice 'magic' and also update the project version automatically, but it would require a bit of setup and some standardization, like:

- Check the branch name and:
- If it is a 'dev/*' branch, update the minor version
- If it is a 'bug/*' branch, update the patch version
- Manual triggers for major and 'non standard' version (e.g. SNAPSHOT) 

...we can think about this when/if we'll have time :D

Reply:
> ... I suppose you get the problem because you have to rebuild the application to generate a fresh .jar with the new version (in this case the issue would be caused by the '`-SNAPSHOT`' after the version that doesn't get recognized by the pseudo-regex I assume)

I thought that docker generates fresh `.jar` every time I build it, isn't that correct? Did you try to run `docker build` on the current version of dockerfile? 

It's just strange that only my machine reproduces this kind of error, I thought that docker should work similarly on every machine.

And I think that the `-SNAPSHOT` is actually recognized by the last `*`, because the wildcard in regex is greedy, so the last `*` matches the whole `1-SNAPSHOT`. Or am I mistaken?

Reply:
I have one last idea, how to fix this:

How about adding this to the `pom.xml`?
```
    <build>
        <plugins>
        ...
        </plugins>
        <finalName>spring-boot</finalName>
    </build>
```
It will force maven to create the final `.jar` with the name `spring-boot.jar`, so we don't have to worry about renaming it using the wildcard pattern during the `COPY`.

This will also resolve my issue with the `docker build`.

Reply:
> > ... I suppose you get the problem because you have to rebuild the application to generate a fresh .jar with the new version (in this case the issue would be caused by the '`-SNAPSHOT`' after the version that doesn't get recognized by the pseudo-regex I assume)
> 
> I thought that docker generates fresh `.jar` every time I build it, isn't that correct? Did you try to run `docker build` on the current version of dockerfile?

True, forgot about that :D 

> It's just strange that only my machine reproduces this kind of error, I thought that docker should work similarly on every machine.
> 
> And I think that the `-SNAPSHOT` is actually recognized by the last `*`, because the wildcard in regex is greedy, so the last `*` matches the whole `1-SNAPSHOT`. Or am I mistaken?

It should, but I'm not really sure in this case. I wrote 'pseudo-regex' because what is actually used is the [Go Match operator](https://pkg.go.dev/path/filepath#Match), and I don't really know how it works. It should be similar to regex, but I've not read into the docs much.

Reply:
Yes let's use something more descriptive, I used `spring-boot` just as an example.

Reply:
Ok, I implemented the last proposal. Thank you!